File: easy/112. Path Sum/112. Path Sum.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        def Run(root, target):
            if not root:
                return False

            if not root.left and not root.right:
                return target - root.val == 0

            return Run(root.left, target) and Run(root.right, target)

        return Run(root, targetSum)

File: easy/112. Path Sum/112. Path Sum.ts
/**
 * Definition for a binary tree node.
 */

class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

function hasPathSum(root: TreeNode | null, targetSum: number): boolean {
    const run = (tree: TreeNode | null, target: number) => {
        if (!tree)
            return false

        if (!tree.right && !tree.left)
            return target - tree.val == 0

        return run(tree.left, target - tree.val) || run(tree.right, target - tree.val)
    }

    return run(root, targetSum)
};

export { }